src/Groups/matrices/FiniteFormOrthogonalGroup.jl

```diff
@@ -798,38 +798,135 @@ function _compute_gens(T::TorQuadMod)
-  (is_degenerate(T) && is_prime(p)) || error("")
-  n = length(invs)
-  q = gram_matrix_quadratic(N)
-  k = count(i -> q[:,i]==0, 1:n)
-  r = n - k
-  Idk = identity_matrix(ZZ, k)
-  Idr = identity_matrix(ZZ, r)
-  NR, i = sub(N, gens(N)[k:end])
-  gensG = fmpz_mat[]
-  for g in _compute_gens(NR)
-    push!(gensG, diagonal_matrix([Idk, g]))
+# We carry the case when T is not semi-regular, but does split its
+# radical quadratic.
+#
+# TODO: We have way too many generators, in both functions. There might be a way
+# to reduce the number of generators along the way!
+function _compute_gens_split_degenerate(T::TorQuadMod)
+  # we need a splitting T = rd + N, where rd is the radical
+  # quadratic and N is a normal form.
+  #
+  # If it is the case, since T splits as the sum of its primary parts,
+  # then each primary part splits its radical quadratic too.
+  # So, we split T into primary parts, since they don't "talk"
+  # to each others via isometries, we compute the orthogonal group
+  # of each primary part and we then "glue" the orthogonal groups
+  # using diagonal matrices on an isometric module to T which is nice enough.
+  @assert has_complement(radical_quadratic(T)[2])[1]
+
+  # if T is "primary degenerate" then we can send it to the other function since
+  # T has only one primary part.
+  if is_prime_power_with_data(elementary_divisors(T)[end])[1]
+    return _compute_gens_split_degenerate_primary(T)
   end
-  if k > 0
-    for g in gens(GL(k,p))
-      push!(gensG, diagonal_matrix([matrix(g), Idr]))
-    end
+
+  # we first create some blocks corresponding to the primary parts of T.
+  # We then compute the orthogonal sum of those blocks, which is a nice
+  # TorQuadMod isometric to T: we get then an isometry between those two modules.
+  gensOT = fmpz_mat[]
+  pd = sort(prime_divisors(order(T)))
+  blocks = TorQuadModMor[primary_part(T, pd[1])[2]]
+  popfirst!(pd)
+  while !isempty(pd)
+    f = blocks[end]
+    ok, j = has_complement(f)
+    @assert ok
+    _T = domain(j)
+    _f = primary_part(_T, pd[1])[2]
+    push!(blocks, compose(_f, j))
+    popfirst!(pd)
   end
-  if k>0 && r>0
-    h = identity_matrix(ZZ, n)
-    for i in 1:r
-      for j in 1:k
-        g = deepcopy(h)
-        g[k+i,j] = 1
-        push!(gensG, g)
-      end
-    end
+  Torth, _, _ = Hecke._orthogonal_sum_with_injections_and_projections(domain.(blocks))
+  ok, phi = is_isometric_with_isometry(Torth, T)
+  @assert ok # Same module with different basis, somehow
+
+  # We compute the orthoognal group of each (primary) block. Since Torth is made
+  # diagonally, we can generate its orthogonal group by taking some diagonal matrices
+  # from taking the cartesian product on the generators of the blocks. We have nothing
+  # to add since we can't map different primary parts to each other.
+  orth_blocks = _compute_gens_split_degenerate_primary.(domain.(blocks))
+  gensOTorth = fmpz_mat[]
+  for x in Hecke.cartesian_product_iterator(orth_blocks, inplace=false)
```

Comment:
This looks like you take the direct sums of all generators. 
But this gives too many generators. Just take `diagonal_matrix(id, id, ... id, f, id, ... id)`
Then you get $sum_i n_i$ generators instead of $\prod_i n_i$ many.